### PR TITLE
DOC: Clarify DeviceStatsMonitor logged metrics (#20807)

### LINKED ISSUE #20807: DeviceStatsMonitor lacks documentation
### 📚 Documentation

I only know of two places where DeviceStatsMonitor() is mentioned in the docs: [here](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.DeviceStatsMonitor.html#lightning.pytorch.callbacks.DeviceStatsMonitor) and [here](https://lightning.ai/docs/pytorch/stable/tuning/profiler_basic.html). Neither place documents what any of the metrics it logs mean!

* e.g. "active.all.current" what does active mean? is this memory or compute? _what units is it in?_
* or "active.large_pool.current" what is the large vs small pool?
* Furthermore [here](https://lightning.ai/docs/pytorch/stable/tuning/profiler_basic.html) it says "ensure that you’re using the full capacity of your accelerator (GPU/TPU/HPU). This can be measured with the DeviceStatsMonitor()" this implies it can measure **GPU utilization** but it does not say which metric records that.

cc @lantiga @borda

### PR DESCRIPTION
**Pull request by MrAnayDongre**
## What does this PR do?

This PR addresses issue #20807 by adding detailed documentation for the metrics logged by `DeviceStatsMonitor`.

The key clarifications include:
- The source of metrics (CPU via `psutil`, CUDA GPU via `torch.cuda.memory_stats`, and other accelerators via `accelerator.get_device_stats()`).
- The naming convention for logged keys: `DeviceStatsMonitor.{hook_name}/{base_metric_name}`.
- Explicitly states that GPU compute utilization is **not** logged by default for CUDA devices, with a pointer to `torch.cuda.memory_stats()` for the full list of memory metrics.
- Provides examples for common CPU, CUDA, and other accelerator (TPU, MPS) metrics.
- Includes a minor update to `profiler_basic.rst` to align with these clarifications and link to the API docs.

This documentation aims to help users understand what statistics to expect when using `DeviceStatsMonitor` with different hardware configurations.

Fixes #20807

No breaking changes are introduced by this documentation update.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs) 
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together? 
- Did you make sure to **update the documentation** with your changes? (if necessary) 
- Did you write any **new necessary tests**? (not for typos and docs) 
- [ ] Did you verify new and **existing tests pass** locally with your changes? 
- Did you list all the **breaking changes** introduced by this pull request? 
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors) 

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/callbacks/device_stats_monitor.py [hunk 1]
```diff
@@ -45,6 +45,23 @@ class DeviceStatsMonitor(Callback):
         ModuleNotFoundError:
             If ``psutil`` is not installed and CPU stats are monitored.
 
+    Logged Metrics:
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/device_stats_monitor.py]: `Raises:` or `Args:` are Sphinx-specific keywords compared to `Logged Metrics:`, so pls let's move it just to the top of this docstring

Comment (MrAnayDongre) [reply src/lightning/pytorch/callbacks/device_stats_monitor.py]: Thanks, @Borda! I've moved the 'Logged Metrics' section to the top of the docstring as requested in commit dcd1042.

Comment (Borda) [reply src/lightning/pytorch/callbacks/device_stats_monitor.py]: well, I still see it without change

Comment (MrAnayDongre) [reply src/lightning/pytorch/callbacks/device_stats_monitor.py]: Hi @Borda,
I've pushed an update (the latest commit on the PR branch is now 2f1c083).
The other CI checks, including the main documentation builds, are now running or queued. Could you please take another look when you have a moment?
Thanks!

File: src/lightning/pytorch/callbacks/device_stats_monitor.py [hunk 2]
```diff
@@ -34,6 +34,22 @@ class DeviceStatsMonitor(Callback):
...
+        All are percentages (%).
+        CUDA GPU (via :func:`torch.cuda.memory_stats`): Logs detailed memory statistics from
+        PyTorch's allocator (e.g., ``allocated_bytes.all.current``, ``num_ooms``; all in Bytes).
+        GPU compute utilization is not logged by default.
+        Other Accelerators (e.g., TPU, MPS): Logs device-specific stats:
+
+        - TPU example: ``avg. free memory (MB)``.
+        - MPS example: ``mps.current_allocated_bytes``.
+
+        Observe logs or check accelerator documentation for details.
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/device_stats_monitor.py]: 'Let''s make this a complete list and you can validate the compiled docs in readthedocs link 📚: [pytorch-lightning--20895.org.readthedocs.build/en/20895](https://pytorch-lightning--20895.org.readthedocs.build/en/20895/)

Comment (MrAnayDongre) [reply src/lightning/pytorch/callbacks/device_stats_monitor.py]: Hi @Borda,
Thanks for your patience and guidance on this!
I've pushed a new set of updates (the latest commit on the branch is df20e4b after pre-commit.ci's auto-fixes).

File: src/lightning/pytorch/callbacks/device_stats_monitor.py [hunk 3]
```diff
@@ -34,6 +34,68 @@ class DeviceStatsMonitor(Callback):
     r"""Automatically monitors and logs device stats during training, validation and testing stage.
     ``DeviceStatsMonitor`` is a special callback as it requires a ``logger`` to passed as argument to the ``Trainer``.
 
+    **Logged Metrics**
+
+    Logs device statistics with keys prefixed as ``DeviceStatsMonitor.{hook_name}/{base_metric_name}``.
+
+    The actual metrics depend on the active accelerator and the ``cpu_stats`` flag.
+
+    **CPU (via `psutil`)**
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/device_stats_monitor.py]: ** really does not well represent the hierarchy in the headers, so let's rather use a nested lists

**Review by SkafteNicki — APPROVED** [review]

File: src/lightning/pytorch/callbacks/device_stats_monitor.py [hunk 4]
```diff
@@ -34,6 +34,67 @@ class DeviceStatsMonitor(Callback):
...
+
+        - Allocator Events:
+
+            - ``num_ooms`` — Cumulative out-of-memory errors
+            - ``num_alloc_retries`` — Number of allocation retries
+            - ``num_device_alloc`` — Number of device allocations
+            - ``num_device_free`` — Number of device deallocations
+
+        For a full list of CUDA memory stats, see the
+        `PyTorch documentation <https://pytorch.org/docs/stable/generated/torch.cuda.memory_stats.html>`_.
```
Comment (Borda) [inline src/lightning/pytorch/callbacks/device_stats_monitor.py]:
```suggestion
        `PyTorch documentation <https://docs.pytorch.org/docs/stable//generated/torch.cuda.device_memory_used.html>`_.
```
